prov/rxd/src/rxd.h

```diff
@@ -205,146 +165,93 @@ static inline struct rxd_cq *rxd_ep_rx_cq(struct rxd_ep *ep)
 	return container_of(ep->util_ep.rx_cq, struct rxd_cq, util_cq);
 }
 
-struct rxd_rx_buf {
-	struct fi_context context;
-	struct slist_entry entry;
-	struct rxd_ep *ep;
-	struct fid_mr *mr;
-	char buf[];
+enum rxd_msg_type {
+	RXD_RTS,
+	RXD_CTS,
+	RXD_ACK,
+	RXD_FREE,
 };
 
-struct rxd_rx_entry {
-	struct ofi_op_hdr op_hdr;
-	uint32_t exp_seg_no;
-	uint64_t msg_id;
-	uint64_t key;
-	uint64_t done;
-	uint64_t peer;
-	uint16_t credits;
-	uint32_t last_win_seg;
-	fi_addr_t source;
-	struct rxd_peer *peer_info;
-	struct rxd_rx_buf *unexp_buf;
-	uint64_t nack_stamp;
-	struct dlist_entry entry;
-
-	union {
-		struct rxd_recv_entry *recv;
-		struct rxd_trecv_entry *trecv;
-
-		struct {
-			struct iovec iov[RXD_IOV_LIMIT];
-		} write;
-
-		struct {
-			struct rxd_tx_entry *tx_entry;
-		} read_rsp;
-	};
-
-	union {
-		struct dlist_entry wait_entry;
-		struct dlist_entry unexp_entry;
-	};
-};
-DECLARE_FREESTACK(struct rxd_rx_entry, rxd_rx_entry_fs);
-
-struct rxd_tx_entry {
+struct rxd_x_entry {
 	fi_addr_t peer;
-	uint64_t msg_id;
-	uint64_t flags;
-	uint64_t rx_key;
-	uint64_t bytes_sent;
+	fi_addr_t peer_x_addr;
+	uint32_t tx_id;
+	uint32_t rx_id;
+	uint32_t key;
+	void *context;
+	enum rxd_msg_type state;
+	uint64_t done;
 	uint32_t seg_no;
 	uint32_t window;
+	uint32_t next_start;
 	uint64_t retry_time;
 	uint8_t retry_cnt;
 
 	struct dlist_entry entry;
 	struct dlist_entry pkt_list;
 
-	uint8_t op_type;
 	struct ofi_op_hdr op_hdr;
+	struct iovec iov[RXD_IOV_LIMIT];
+	uint8_t iov_count;
+};
+DECLARE_FREESTACK(struct rxd_x_entry, rxd_tx_fs);
+DECLARE_FREESTACK(struct rxd_x_entry, rxd_rx_fs);
 
-	union {
-		struct {
-			struct fi_msg msg;
-			struct iovec msg_iov[RXD_IOV_LIMIT];
-		} msg;
+#define RXD_RETRY	(1 << 0)
+#define RXD_LAST	(1 << 1)
 
-		struct {
-			struct fi_msg_tagged tmsg;
-			struct iovec msg_iov[RXD_IOV_LIMIT];
-		} tmsg;
+struct rxd_data_hdr {
+	uint32_t tx_id;
+	uint32_t rx_id;
+	uint32_t key;
+	uint32_t seg_no;
+	uint32_t flags;
+	fi_addr_t peer;
+};
 
-		struct {
-			struct fi_msg_rma msg;
-			struct iovec src_iov[RXD_IOV_LIMIT];
-			struct fi_rma_iov dst_iov[RXD_IOV_LIMIT];
-		} write;
+struct rxd_ctrl_hdr {
+	uint32_t tx_id;
+	uint32_t rx_id;
+	uint32_t key;
 
-		struct {
-			struct fi_msg_rma msg;
-			struct fi_rma_iov src_iov[RXD_IOV_LIMIT];
-			struct iovec dst_iov[RXD_IOV_LIMIT];
-		} read_req;
+	fi_addr_t peer;
+	uint8_t flags;
+	enum rxd_msg_type type;
+	uint32_t window;
+	size_t size;
+	uint64_t data;
+};
 
+struct rxd_pkt {
+	union {
+		struct {
+			uint32_t resv;//always = ~0
+			struct rxd_ctrl_hdr ctrl_hdr;
+			uint8_t source[];
+		};
 		struct {
-			uint64_t peer_msg_id;
-			uint8_t iov_count;
-			struct iovec src_iov[RXD_IOV_LIMIT];
-		} read_rsp;
+			struct rxd_data_hdr data_hdr;
+			uint8_t data[];
+		};
```

Comment:
There needs to be a common header for all packets.  That header should have a version field, plus any fields that are needed for all/most packets, including a type field to distinguish between data versus ctrl packets.  We need to ensure that protocol changes can be made without breaking communication with peers that may not have the same updates.  That is, you can't assume that the peer is running the same version of the library.